modules/stochastic_tools/src/surrogates/LibtorchSimpleNNTrainer.C

```diff
@@ -0,0 +1,199 @@
+{
+  unsigned int num_parameters = _sampler.getNumberOfCols();
+
+  for (unsigned int i = 0; i < num_parameters; ++i)
+    _flattened_data[_local_row * num_parameters + i] = _sampler_row[i];
+
+  _flattened_response[_local_row] = _response;
+}
+
+void
+LibtorchSimpleNNTrainer::postTrain()
+{
+  _communicator.allgather(_flattened_data);
+  _communicator.allgather(_flattened_response);
+
+#ifdef LIBTORCH_ENABLED
+
+  // Then, we create and load our Tensors
+  unsigned int n_rows = _flattened_response.size();
+  unsigned int n_cols = _sampler.getNumberOfCols();
+
+  // The default data type in pytorch is float, while we use double in MOOSE.
+  // Therefore, in some cases we have to convert Tensors to double.
+  auto options = torch::TensorOptions().dtype(at::kDouble);
+  torch::Tensor data_tensor =
+      torch::from_blob(_flattened_data.data(), {n_rows, n_cols}, options).to(at::kDouble);
+  torch::Tensor response_tensor =
+      torch::from_blob(_flattened_response.data(), {n_rows}, options).to(at::kDouble);
+
+  // We create a custom data loader which can be used to select samples for the in
+  // the training process. See the header file for the definition of this structure.
+  StochasticTools::LibtorchDataset my_data(data_tensor, response_tensor);
+
+  // We initialize a data_loader for the training part.
+  unsigned int sample_per_batch = n_rows > _num_batches ? n_rows / _num_batches : 1;
+
+  auto data_set = my_data.map(torch::data::transforms::Stack<>());
+  auto data_loader = torch::data::make_data_loader<torch::data::samplers::SequentialSampler>(
+      std::move(data_set), sample_per_batch);
+
+  // We create a neural net (for the definition of the net see the header file)
+  _nn = std::make_shared<StochasticTools::LibtorchSimpleNeuralNet>(
+      _filename, n_cols, _num_neurons_per_layer, 1);
+
+  // Initialize the optimizer
+  torch::optim::Adam optimizer(_nn->parameters(), torch::optim::AdamOptions(_learning_rate));
+
+  if (_read_from_file)
+    try
+    {
+      torch::load(_nn, _filename);
+      _console << "Loaded requested .pt file." << std::endl;
+    }
+    catch (...)
+    {
+      mooseError("The requested pytorch file could not be loaded.");
+    }
+
+  Real rel_loss = 1.0;
+  Real initial_loss = 1.0;
+  Real epoch_loss = 0.0;
+
+  // Begin training loop
+  unsigned int epoch = 1;
+  while (epoch <= _num_epocs && rel_loss > _rel_loss_tol)
+  {
+    epoch_loss = 0.0;
+    for (auto & batch : *data_loader)
+    {
+      // Reset gradients
+      optimizer.zero_grad();
+
+      // Compute prediction
+      torch::Tensor prediction = _nn->forward(batch.data);
+
+      // Compute loss values using a MSE ( mean squared error)
+      torch::Tensor loss = torch::mse_loss(prediction.reshape({prediction.size(0)}), batch.target);
+
+      // Propagate error back
+      loss.backward();
+
+      // Use new gradients to update the parameters
+      optimizer.step();
+
+      epoch_loss += loss.item<double>();
+    }
```

Comment:
Curious why the actual calculations (training) happens in `posttrain` not `train`

Reply:
Because we have the following training procedure:
```
preTrain()
for (samples in the sampler)
   train()
postTrain()
```

So we only want to optimize the neural net once every data is available to us.